internal/app/machined/pkg/runtime/v1alpha1/platform/hcloud/hcloud.go

```diff
@@ -0,0 +1,237 @@
+	"github.com/talos-systems/talos/pkg/download"
+	"github.com/talos-systems/talos/pkg/machinery/config"
+	"github.com/talos-systems/talos/pkg/machinery/config/configloader"
+	"github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1"
+)
+
+const (
+	// HCloudExternalIPEndpoint is the local hcloud endpoint for the external IP.
+	HCloudExternalIPEndpoint = "http://169.254.169.254/hetzner/v1/metadata/public-ipv4"
+
+	// HCloudNetworkEndpoint is the local hcloud endpoint for the network-config.
+	HCloudNetworkEndpoint = "http://169.254.169.254/hetzner/v1/metadata/network-config"
+
+	// HCloudHostnameEndpoint is the local hcloud endpoint for the hostname.
+	HCloudHostnameEndpoint = "http://169.254.169.254/hetzner/v1/metadata/hostname"
+
+	// HCloudUserDataEndpoint is the local hcloud endpoint for the config.
+	HCloudUserDataEndpoint = "http://169.254.169.254/hetzner/v1/userdata"
+)
+
+// NetworkConfig holds hcloud network-config info.
+type NetworkConfig struct {
+	Version int       `yaml:"version"`
+	Config  []Network `yaml:"config"`
+}
+
+// Network holds network info from the hcloud network-config metadata.
+type Network struct {
+	Mac        string    `yaml:"mac_address"`
+	Interfaces string    `yaml:"name"`
+	Subnets    []Subnets `yaml:"subnets"`
+	Type       string    `yaml:"type"`
+}
+
+// Subnets holds network info from the hcloud network-config subnets metadata.
+type Subnets struct {
+	NameServers []string `yaml:"dns_nameservers"`
+	Address     string   `yaml:"address,omitempty"`
+	Gateway     string   `yaml:"gateway,omitempty"`
+	Ipv4        bool     `yaml:"ipv4,omitempty"`
+	Ipv6        bool     `yaml:"ipv6,omitempty"`
+	Type        string   `yaml:"type"`
+}
+
+// Hcloud is the concrete type that implements the runtime.Platform interface.
+type Hcloud struct{}
+
+// Name implements the runtime.Platform interface.
+func (h *Hcloud) Name() string {
+	return "hcloud"
+}
+
+// ConfigurationNetwork implements the network configuration interface.
+//nolint:gocyclo
+func (h *Hcloud) ConfigurationNetwork(metadataNetworkConfig []byte, confProvider config.Provider) (config.Provider, error) {
+	var unmarshalledNetworkConfig NetworkConfig
+
+	if err := yaml.Unmarshal(metadataNetworkConfig, &unmarshalledNetworkConfig); err != nil {
+		return nil, err
+	}
+
+	if unmarshalledNetworkConfig.Version != 1 {
+		return nil, fmt.Errorf("network-config metadata version=%d is not support", unmarshalledNetworkConfig.Version)
+	}
+
+	var machineConfig *v1alpha1.Config
+
+	machineConfig, ok := confProvider.(*v1alpha1.Config)
+	if !ok {
+		return nil, fmt.Errorf("unable to determine machine config type")
+	}
+
+	if machineConfig.MachineConfig == nil {
+		machineConfig.MachineConfig = &v1alpha1.MachineConfig{}
+	}
+
+	if machineConfig.MachineConfig.MachineNetwork == nil {
+		machineConfig.MachineConfig.MachineNetwork = &v1alpha1.NetworkConfig{
+			NetworkInterfaces: []*v1alpha1.Device{
```

Comment:
as this is the default (unless `eth0` has any other configuration), do we even need to add it? is the plan to have both DHCP and static config? (this can be configured on single interface entry now)

Reply:
DHCP for IP-v4, static for IP-v6

Ok, I'll try to merge it

Reply:
ah, ok, then we can have that in the single `interface` section, that's 0.12 feature :)